Memory routines refactor

rr- opened the pull request:
I'm not sure if I like the new wordy enum constants, but here goes.

walkawayy commented:
I like the new structure. If you think the new enum names are too long, you could shorten the name a little with `MEM_BUF_TEXTURE_PAGES ` etc. Cuts off 6 characters?

oziphantom commented:
I feel BUFFER is redundant, its not really a buffer as its the whole games memory, but it is stored in a buffer, but since all of it is, it's redundant. just `MEMORY_XXXXXX` which you could Hungarian it. `memTEXTURE_PAGES`

oziphantom commented on src/specific/memory.c:
```diff
@@ -9,7 +9,6 @@
 
 static char *m_GameMemoryPointer = NULL;
 static char *m_GameAllocMemPointer = NULL;
-static uint32_t m_GameAllocMemUsed = 0;
```
I feel a better use of this is to keep it and throw asserts or a S_ExitSystemFmt on it, so we actually check we don't run out of RAM. While at 16MB no original level is going to, but TRLE mappers be crazy.

rr- replied:
We can infer this value directly from m_GameAllocMemFree + we already check for memory exhaustion.

oziphantom replied:
yeah, missed that. I guess it was originally for a screen print.

rr- commented:
> I like the new structure. If you think the new enum names are too long, you could shorten the name a little with `MEM_BUF_TEXTURE_PAGES` etc. Cuts off 6 characters?

I'll go with that

rr- commented:
Done

thatguyfrombb commented:
Writing it here as something to check on later: `Alloc` uses adjusted sizes (padded to however many bytes). Does `free` use adjusted sizes or direct (single byte) sizes?

rr- commented:
@thatguyfrombb I wouldn't worry too much about it, the implementation of `Memory_Free` is dumb anyway and frees only last few bytes.

oziphantom commented:
It's used once in the whole game, and that allocs in bytes and does zero fix up. 

```
char **sample_pointers =
        game_malloc(sizeof(char *) * num_samples, GBUF_SAMPLE_OFFSETS);
    for (int i = 0; i < num_samples; i++) {
        sample_pointers[i] = sample_data + sample_offsets[i];
    }

    SoundLoadSamples(sample_pointers, num_samples);

    game_free(sizeof(char *) * num_samples, GBUF_SAMPLE_OFFSETS);
```

rr- commented:
In fact I'm wondering whether if we shouldn't just go with `malloc()` for this module. The only allocations happen near FMVs and level loading so it's not like we'll get hurt by the performance. Then we can keep a pointer map and call free() in Memory_Free() like a normal human being.

Furthermore we already have a memory leak since each LoadLevel calls Memory_Init but nobody calls Memory_Shutdown, so we end up allocating 16 mb more every level.

rr- commented:
Like here, in the latest commit ^
IMO it looks much nicer and is less hacky.

oziphantom commented:
that has to be a bug, there is no way this game would get past 1 level on a PS1 or Saturn if that was the case and back in the day it would make it past 4 levels on a PC of the era. 
It solved a very real problem form the era, which we don't have anymore. But still seems like adding extra overhead when we don't need it?

oziphantom commented:
You're new method only keeps the last alloc for each type, not all allocs for each type. 

```
    FileRead(&MeshPtrCount, sizeof(int32_t), 1, fp);
    uint32_t *mesh_indices =
        game_malloc(sizeof(uint32_t) * MeshPtrCount, GBUF_MESH_POINTERS);
    FileRead(mesh_indices, sizeof(uint32_t), MeshPtrCount, fp);

    Meshes = game_malloc(sizeof(int16_t *) * MeshPtrCount, GBUF_MESH_POINTERS);
    for (int i = 0; i < MeshPtrCount; i++) {
        Meshes[i] = &MeshBase[mesh_indices[i] / 2];
    }
```
in this case the first `mesh_indicies` alloc is lost.

rr- commented:
True, didn't think of it. I'll fix this.

rr- commented:
I tried to go at my on-the-fly allocation idea from multiple angles, but there was always something: for example we expect to grow the memory in the LoadRooms, but after I changed that routine to loop over the file twice and allocate things just once, it turned out that there are items that use Memory_Alloc, which means I can't allocate the data just once for them in a neat way. And realloc is also no go because it invalidates the old pointers other modules keep around.

So I discarded this idea and instead removed Memory_Free completely. In the places that can and should free their memory, I added a small wrapper specific/memory.c and .h that checks if we succeeded to allocate the needed memory, and exits otherwise.

I'm now thinking of the naming. Now that I think of it, Memory for the game buffer is misleading, because it really is related to just the game structures, not general memory management. Maybe it should be GameBuffer after all.
And specific/memory.c is not really platform-specific code. Maybe it could become just memory.c. What do you think?
@oziphantom @thatguyfrombb

oziphantom commented on src/specific/file.c:
```diff
@@ -575,19 +576,18 @@ static bool LoadSamples(MYFILE *fp)
 
-    char **sample_pointers =
-        Memory_Alloc(sizeof(char *) * num_samples, MEM_BUF_SAMPLE_OFFSETS);
+    char **sample_pointers = S_Memory_Alloc(sizeof(char *) * num_samples);
```
I feel this is now overkill as we have c99 and this only exists in this scope as they are removed bellow. We could just do `int32 sample_offsets[num_samples];` and likewise for the pointers and just let them fall out of scope. The number of samples is probably below 256 to which adding 2K onto the stack is nothing these days.

rr- commented:
I don't like declaring arrays of unknown sizes on the stack, so I'm against this.

On Fri, Nov 12, 2021 at 04:14:56AM -0800, oziphantom wrote:
> @oziphantom commented on this pull request.
> 
> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
> 
> In src/specific/file.c:
> 
> >      FileRead(sample_offsets, sizeof(int32_t), num_samples, fp);
> 
> -    char **sample_pointers =
> -        Memory_Alloc(sizeof(char *) * num_samples, MEM_BUF_SAMPLE_OFFSETS);
> +    char **sample_pointers = S_Memory_Alloc(sizeof(char *) * num_samples);
> 
> I feel this is now overkill as we have c99 and this only exists in this scope as they are removed bellow. We could just do int32 sample_offsets[num_samples];
> and likewise for the pointers and just let them fall out of scope. The number of samples is probably below 256 to which adding 2K onto the stack is nothing
> these days.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. *
>

oziphantom commented:
as to being 'specific', as we will not be supporting the Saturn or PS1 all the memory allocations are no longer platform specific, thus I agree.

rr- commented:
Alright I think I'm fairly satisfied with this PR now, LMK your thoughts :) 
Sorry for pushing it in half-WIP state, some of the ideas didn't occur to me until after I requested the initial reviews.

oziphantom reviewed (commented):
this had pending too

rr- commented on src/specific/sndpc.c:
```diff
@@ -127,7 +127,8 @@ void *SoundPlaySample(
 
-            DUPE_SOUND_BUFFER *dupe_buffer = malloc(sizeof(DUPE_SOUND_BUFFER));
+            DUPE_SOUND_BUFFER *dupe_buffer =
+                S_Memory_Alloc(sizeof(DUPE_SOUND_BUFFER));
```
They're never freed, they last for the lifetime of the game execution, but that should be fine since they will be reused all the time.

rr- commented on src/game/draw.c:
```diff
@@ -448,9 +446,9 @@ void DrawPickupItem(ITEM_INFO *item)
+            int centred = ABS(min_y + max_y) < 8;
             if (floor_mapped_delta) {
-                offset = item->pos.y - abs(min_y - sprite->y1) / 8;
+                offset = item->pos.y - ABS(min_y - sprite->y1) / 8;
```
Good point, I'll split this into a separate PR

rr- commented on src/specific/sndpc.c:
```diff
@@ -253,7 +254,7 @@ SAMPLE_DATA *SoundLoadSample(char *content)
 
-    SAMPLE_DATA *sample_data = malloc(sizeof(SAMPLE_DATA));
+    SAMPLE_DATA *sample_data = S_Memory_Alloc(sizeof(SAMPLE_DATA));
     memset(sample_data, 0, sizeof(SAMPLE_DATA));
```
Good catch

rr- replied:
Fixed

thatguyfrombb reviewed (changes requested):
Testing went good :)

rr- commented on src/memory.c:
```diff
@@ -0,0 +1,32 @@
+void Memory_Free(void *memory)
+{
+    if (memory) {
+        free(memory);
```
Note that this assignment will only work in the scope of the function, so it won't affect the callers of this function.

rr- commented on src/specific/smain.c:
```diff
@@ -1,6 +1,7 @@
 #include "specific/smain.h"
 
 #include "config.h"
+#include "game/gamebuf.h"
```
Fixed

rr- commented on src/specific/display.c:
```diff
@@ -1,14 +1,12 @@
 #include "specific/display.h"
 
 #include "3dsystem/3d_gen.h"
+#include "game/gamebuf.h"
```
Fixed

rr- commented on src/game/gamebuf.c:
```diff
@@ -0,0 +1,85 @@
+
+    aligned_size = (alloc_size + 3) & ~3;
+
+    if (aligned_size > m_GameAllocMemFree) {
```
Fixed

rr- commented on src/game/settings.c:
```diff
@@ -151,7 +151,7 @@ static int32_t S_ReadUserSettingsT1M()
     cfg_data_size = FileSize(fp);
-    cfg_data = malloc(cfg_data_size + 1);
+    cfg_data = Memory_Alloc(cfg_data_size + 1);
     if (!cfg_data) {
```
Fixed

rr- commented on src/config.c:
```diff
@@ -194,12 +194,7 @@ int8_t T1MReadConfig()
-        result = T1MReadConfigFromJson("");
-        goto cleanup;
-    }
+    cfg_data = Memory_Alloc(cfg_data_size + 1);
```
Fixed
